source/slang/ir-serialize.cpp

```diff
@@ -633,79 +727,509 @@ Result _writeArrayChunk(uint32_t chunkId, const List<T>& array, Stream* stream)
+    return _readArrayChunk(Bin::CompressionType::None, chunk, stream, numReadInOut, resizer);
+}
+
+static Result _decodeInsts(IRSerialBinary::CompressionType compressionType, const List<uint8_t>& encodeIn, List<IRSerialData::Inst>& instsOut)
+{
+    typedef IRSerialBinary Bin;
+    typedef IRSerialData::Inst::PayloadType PayloadType;
+
+    if (compressionType != Bin::CompressionType::VariableByteLite)
     {
         return SLANG_FAIL;
     }
 
-    arrayOut.SetSize(header.m_numEntries);
+    const size_t numInsts = size_t(instsOut.Count());
+    IRSerialData::Inst* insts = instsOut.begin();
+
+    const uint8_t* encodeCur = encodeIn.begin();
+
+    for (size_t i = 0; i < numInsts; ++i)
+    {
+        auto& inst = insts[i];
+
+        inst.m_op = *encodeCur++;
+        const PayloadType payloadType = PayloadType(*encodeCur++);
+        inst.m_payloadType = payloadType;
+        
+        // Read the result value
+        encodeCur += ByteEncodeUtil::decodeLiteUInt32(encodeCur, (uint32_t*)&inst.m_resultTypeIndex);
+
+        switch (inst.m_payloadType)
+        {
+            case PayloadType::Empty:
+            {
+                break;
+            }
+            case PayloadType::Operand_1:
+            case PayloadType::String_1:
+            case PayloadType::UInt32:
+            {
+                // 1 UInt32
+                encodeCur += ByteEncodeUtil::decodeLiteUInt32(encodeCur, (uint32_t*)&inst.m_payload.m_operands[0]);
+                break;
+            }
+            case PayloadType::Operand_2:
+            case PayloadType::OperandAndUInt32:
+            case PayloadType::OperandExternal:
+            case PayloadType::String_2:
+            {
+                // 2 UInt32
+                encodeCur += ByteEncodeUtil::decodeLiteUInt32(encodeCur, 2, (uint32_t*)&inst.m_payload.m_operands[0]);
+                break;
+            }
+            case PayloadType::Float64:
+            {
+                memcpy(&inst.m_payload.m_float64, encodeCur, sizeof(inst.m_payload.m_float64));
+                encodeCur += sizeof(inst.m_payload.m_float64);
+                break;
+            }
+            case PayloadType::Int64:
+            {
+                memcpy(&inst.m_payload.m_int64, encodeCur, sizeof(inst.m_payload.m_int64));
+                encodeCur += sizeof(inst.m_payload.m_int64);
+                break;
+            }
+        }
+    }
+
+    return SLANG_OK;
+}
 
-    stream->Read(arrayOut.begin(), sizeof(T) * header.m_numEntries);
+Result _readInstArrayChunk(const IRSerialBinary::SlangHeader& slangHeader, const IRSerialBinary::Chunk& chunk, Stream* stream, size_t* numReadInOut, List<IRSerialData::Inst>& arrayOut)
+{
+    typedef IRSerialBinary Bin;
+
+    Bin::CompressionType compressionType = Bin::CompressionType::None;
+    if (chunk.m_type == SLANG_MAKE_COMPRESSED_FOUR_CC(chunk.m_type))
+    {
+        compressionType = Bin::CompressionType(slangHeader.m_compressionType);
+    }
+
+    switch (compressionType)
+    {
```

Comment:
It seems to me like the combination of (1) making the bytestream compaction an optional feature and (2) writing the serialization in this "struct-oriented" fashion instead of something "stream-oriented" means that there is a lot of code that needs to carefully branch to handle all the cases.

E.g., a purely "stream-oriented" style of serialization could easily abstract over whether the stream is byte-compacted by hiding the branch inside the stream reader/writer type (so that the logic that read/writes a particular kind of object on the stream doesn't have to care about the encoding mode).

The main reason that stream-oriented approach doesn't work here seems to be the desire to have the uncompressed "on disk" format be 1-to-1 with the "second format" in memory.